### PR TITLE
Fix music and SFX on mobile

### PR DESCRIPTION
**Pull request by islemaster**
Fixes a set of issues related to audio problems on mobile devices.
- On mobile devices, web audio is typically "locked" until a sound is triggered directly by a user interaction.  This makes autoplaying music difficult (or impossible).  As a result, we now try starting our music when the instructions dialog is dismissed (if it wasn't played when they were shown).
- We now play a brief silent sound on the first user interaction in order to unlock web audio / verify that web audio is unlocked.  I also added a way to unhook a click/touch handler added via `dom.js` since only only need it once (how did this not exist before?).
- A new limitation in iOS9 is that you can't unlock audio on a `touchstart` event, so we use a `touchend` event to unlock audio now.
- We had a regression related to skipping videos because local storage told us to - if we are on a level with an autoplay video but we skip the video, we would skip the instructions too.  This makes it so we are back to always showing instructions.  Skipping instruction was making us also skip the music, since that was our trigger for starting music. I'll submit a UI test for this soon.

Tested via star wars tutorial on:
- Desktop Chrome
- IE11
- iOS7 Safari & Chrome
- iOS8.1 Safari & Chrome
- iOS9.1 Safari
- Android 4.4 Chrome

Key resources:
- https://paulbakaus.com/tutorials/html5/web-audio-on-ios/
- http://www.holovaty.com/writing/ios9-web-audio/

### CODE REVIEW

**Comment by islemaster** [comment]
cc @mehalshah Please review the fix for the instructions regression (see videos.js).  This is the second regression I've found related to the local storage stuff - in both cases skipping a video caused us to skip other behaviors we weren't supposed to skip.  Any thoughts on how we can track down any remaining regressions around that change?

**Comment by islemaster** [comment]
cc @philbogle HoC-related audio fixes.

**Comment by islemaster** [comment]
cc @cpirich Audio fixes requiring some invasive stuff around touch events, so more eyes on this will be better.

File: dashboard/app/assets/javascripts/soundEffects.js [hunk 1]
```diff
@@ -67,6 +75,45 @@ function Sounds() {
...
+  return this.audioUnlocked_ || !this.audioContext;
+};
+
+/**
+ * Mobile browsers disable audio until a sound is triggered by user interaction.
+ * This method tries to play a brief silent clip to test whether audio is
+ * unlocked, and/or trigger an unlock if called inside a user interaction.
+ */
+Sounds.prototype.unlockAudio = function () {
+  if (this.audioUnlocked_ || !this.audioContext) {
```
Comment (islemaster) [inline dashboard/app/assets/javascripts/soundEffects.js]: I should call `this.isAudioUnlocked()` here.  Fixing...

**Comment by islemaster** [comment]
cc @bcjordan because you've done a bunch of audio stuff too.

**Comment by islemaster** [comment]
Legitimate failures in CircleCI.  Working on it...

**Comment by islemaster** [comment]
Tests fixed.

File: dashboard/app/assets/javascripts/videos.js.erb [hunk 2]
```diff
@@ -23,15 +23,22 @@ function createVideo(options) {
...
   }
 
+  if (options.onClose === undefined) {
+    options.onClose = function () {};
+  }
+
   if (dashboard.clientState.hasSeenVideo(options.key) && forceShowVideo === false) {
     // Anything we were going to do when the video closed, we ought to do
     // right now.
     // TODO: What else are we missing here?
```
Comment (mehalshah) [inline dashboard/app/assets/javascripts/videos.js.erb]: I don't think you are missing anything. This is the only place where we skip showing the video

**Comment by breville** [comment]
lgtm, but glad you've asked all the right people for input.

File: dashboard/app/assets/javascripts/videos.js.erb (hunk 2 above)
Comment (islemaster) [inline dashboard/app/assets/javascripts/videos.js.erb]: Removing the todo

**Pull request merged**